hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -459,6 +459,26 @@ def contains(self, item):
         """
         return self._method("contains", tbool, item)
 
+    def head(self):
+        """Returns the first element of the array, or missing if empty.
+
+        Returns
+        -------
+        :class:`.Expression`
+            Element.
+
+        Examples
+        --------
+        >>> hl.eval(names.head())
+        'Alice'
+
+        If the array has no elements, then the result is missing:
+        >>> hl.eval(names.filter(lambda x: x.startswith('D')).head())
+        None
+        """
+        # FIXME: this should generate short-circuiting IR when that is possible
+        return hl.rbind(self, lambda x: hl.case().when(x.length() > 0, x[0]).or_missing())
```

Comment:
Any reason not to just `hl.or_missing(hl.len(x) > 0, x[0])`

Reply:
I don't like that function, I find it confusing

Reply:
both will build the same IR